### DISCUSSION TITLE
Laravel 11 Route Model Binding: Why the Detailed Error Message Regardless of Environment?

### DISCUSSION THREAD

**Discussion by ozdemirrulass** [laravel/framework#52575] [Q&A]
In Laravel 11, there's something I find confusing and would like clarification on. When using route model binding, a ModelNotFoundException is translated into a NotFoundHttpException.

In addition to that regardless of whether the debug mode is active or not, or if the environment is set to production or local. The error message directly returns the model namespace and identifier, like:

`No query results for model [App\Models\<MODEL_NAME>] <MODEL_IDENTIFIER>`

I used a custom ApiException class and implemented custom behavior in app.php to solve it, but this still seems a bit weird.

Can anyone shed light on why this is the case?

**Comment by henzeb** [comment]
Did you set debug to false as well?

**Reply by ozdemirrulass** [reply]
>"In addition to that regardless of whether the debug mode is active or not"

**Comment by noefleury** [comment]
About the exception which is transformed to a `NotFoundHttpException` I think that’s because from a REST perspective the route is not found (but it’s discutable).
For the second question I’m curious too to know the answer, and more particularly when in `production`.

**Reply by ozdemirrulass** [reply]
I thought the same. I think it would be better to change the behavior of the exceptions based on the environment.

**Comment by Abouelez** [comment]
You can override response when ever this exception is thrown, 
in bootstrap/app.php you can do that
```
 ->withExceptions(function (Exceptions $exceptions) {
       $exceptions->renderable(function (NotFoundHttpException $e, $request) {
            if ($request->is('api/*')) {
                return response()->json(['message' => 'Record not found.'], 404);
            }
            throw $e; //for web view
        });
})
```

**Reply by max-shteklia** [reply]
What about showing at least the model's name? Instead of "No query results for model [App\Models\<MODEL_NAME>] <MODEL_IDENTIFIER>" return custom "The specified <MODEL_NAME> was not found"

**Reply by max-shteklia** [reply]
Just found some article and modified the code to get the model. 
Article says to modify for production for ModelNotFoundException:
```
->withExceptions(function (Exceptions $exceptions) {
        if (app()->isProduction()) {
            $exceptions->map(
                ModelNotFoundException::class,
                function (ModelNotFoundException $e): NotFoundHttpException {
                    return new NotFoundHttpException("Not found", $e);
                }
            );
        }
        ...
```

I changed to show the exact Model name with some modifications for getting human name for model (including multiword model name such as OrderPayment -> Order Payment)
```
->withExceptions(function (Exceptions $exceptions) {
        if (app()->isProduction()) {
            $exceptions->map(
                ModelNotFoundException::class,
                function (ModelNotFoundException $e): NotFoundHttpException {
                    $modelName = Str::headline(class_basename("App\Models\Study"));
                    return new NotFoundHttpException("The specified $modelName was not found", $e);
                }
            );
        }
    })->create();

```

Also, I'm going to create a macro for Str to simplify model name converting:
```
Str::macro(
        'readableModel',
        fn (string $string): string => Str::headline(class_basename($string))
);
```

PS While writing this, I thought that actually `if (app()->isProduction()) {...}` in that exception is optional and I want to remove it in my project

**Reply by Abouelez** [reply]
Ok that's a good approach , but if i use it should i add these lines brlow to ensure  JSON response for API? or it will handle it

```
 ->withExceptions(function (Exceptions $exceptions) {
       $exceptions->renderable(function (NotFoundHttpException $e, $request) {
            if ($request->is('api/*')) {
                return response()->json(['message' => $e->getMessage()], 404);
            
        });
})
```

**Reply by ghost** [reply]
You should say {resourceName} not found not model name not found.

GET /actions/{id}

could use Model Payment but the message should say 'Action not found';

example https://github.com/macropay-solutions/laravel-crud-wizard-free/blob/7d1e1d24e257f51b4249a7ca4d0e09f097bd0d7f/src/Helpers/GeneralHelper.php#L36

**Reply by max-shteklia** [reply]
> Ok that's a good approach , but if i use it should i add these lines brlow to ensure JSON response for API? or it will handle it
> 
> ```
>  ->withExceptions(function (Exceptions $exceptions) {
>        $exceptions->renderable(function (NotFoundHttpException $e, $request) {
>             if ($request->is('api/*')) {
>                 return response()->json(['message' => $e->getMessage()], 404);
>             
>         });
> })
> ```
Well, I haven't tested it as not an api request, since I'm working right now on API only app, however, as I remember, on 404 laravel just returns 404 page, so it might be redundant to add `if ($request->is('api/*')) {...}`. The best thing would be to try it out:)

**Reply by max-shteklia** [reply]
> You should say {resourceName} not found not model name not found.
> 
> GET /actions/{id}
> 
> could use Model Payment but the message should say 'Action not found';

That is a valid point too. In my projects, I mostly use route model binding, so my models are according to error messages I need. What defines `$e->getModel()::RESOURCE_NAME`? I started exploring in the web what this constant is and found nothing; I'm assuming that it is the custom thing you add in your model, am I right? Do you add a `RESOURCE_NAME` constant to every model?

**Reply by ghost** [reply]
@max-shteklia I use that lib for crud and filtering operations
The RESOURCE_NAME is defined in the BaseModel https://github.com/macropay-solutions/laravel-crud-wizard-free/blob/7d1e1d24e257f51b4249a7ca4d0e09f097bd0d7f/src/Models/BaseModel.php#L24

You can have a look at this demo project https://github.com/macropay-solutions/laravel-crud-wizard-demo. I used it when I first started using the lib.

I opened a discussion and suggested them to use Str::singular for the resource name: https://github.com/macropay-solutions/laravel-crud-wizard-free/discussions/9

**Reply by macropay-solutions** [reply]
@max-shteklia you can use this https://laravel.com/docs/11.x/routing#customizing-missing-model-behavior to customize the missing model if using our lib is overhead for you.


Update:
You inspired us to add to the lib also **get update and delete** related by identifier in version 3.4.13 for URI like: 

`/resource/{identifier}/{relation}/{relatedIdentifier}`

example: http://89.40.19.34/laravel-10-free/laravel-lumen-crud-wizard#iii2-get-resource

**Reply by max-shteklia** [reply]
@bulletproof-coding Doesn't that mean you have to manually name the resource in every model? I like the way it shows just the model's name; it works perfectly from my project's perspective and doesn't reveal anything. Your point remains valid; I simply don't want to define the resource name manually for every model because, in my situation, it would still duplicate my model's name.

**Reply by max-shteklia** [reply]
@macropay-solutions Making every route using missing() is a pretty weird approach, in my opinion. Your lib is not overhead for me. I just wanted to change the not found message and found out that in Laravel, there is not a super easy solution for that, and I shared my approach for solving that :)

**Reply by macropay-solutions** [reply]
@max-shteklia @bulletproof-coding Anyway thank you both for inspiring us.

**Reply by ghost** [reply]
@max-shteklia 

> Doesn't that mean you have to manually name the resource in every model?

Yes it does. It helps with auto registering all routes but you have to manually write each resource name in each model you expose.

**Reply by macropay-solutions** [reply]
> > You should say {resourceName} not found not model name not found.
> > GET /actions/{id}
> > could use Model Payment but the message should say 'Action not found';
> 
> That is a valid point too. In my projects, I mostly use route model binding, so my models are according to error messages I need. What defines `$e->getModel()::RESOURCE_NAME`? I started exploring in the web what this constant is and found nothing; I'm assuming that it is the custom thing you add in your model, am I right? Do you add a `RESOURCE_NAME` constant to every model?

@max-shteklia now RESOURCE_NAME is optional just like you wanted. https://github.com/macropay-solutions/laravel-crud-wizard-free/releases/tag/3.4.14
